File: 6-Z字形变换.py
#
# @lc app=leetcode.cn id=6 lang=python3
#
# [6] Z 字形变换
#

# @lc code=start
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        n, r = len(s), numRows
        if r == 1 or r >= n:
            return s
        t = 2 * r - 2
        res = []
        for i in range(r):
            for j in range(0, n-i, t):
                res.append(s[j+i])
                if 0 < i < r-1 and j + t - i < n:
                    res.append(s[j+t-i])
        return "".join(res)


# @lc code=end

s = "PAYPALISHIRING"
numRows = 3
sol = Solution().convert(s, numRows)
print(sol)

File: 6-Z字形变换.cpp
#include <iostream>
using namespace std;
#include <string>

class Solution
{
public:
    string convert(string s, int numRows)
    {
        int n = s.size();
        int r = numRows;
        if (r == 1 || r >= n)
        {
            return s;
        }
        int t = 2 * r - 2;
        string res;
        for (int i = 0; i < r; i++)
        {
            for (int j = 0; j < n - i; j += t)
            {
                res.push_back(s[j + i]);
                if (0 < i && i < r - 1 && j + t - i < n)
                {
                    res.push_back(s[j + t - i]);
                }
            }
        }
        return res;
    }
};

int main()
{
    string s = "PAYPALISHIRING";
    int numRows = 3;
    Solution sol;
    string res = sol.convert(s, numRows);
    cout << res << endl;
}